concent_api/api-integration-multi-threads-test.py

```diff
@@ -0,0 +1,201 @@
+) -> None:
+    api_request(
+        cluster_url,
+        'send',
+        REQUESTOR_PRIVATE_KEY,
+        CONCENT_PUBLIC_KEY,
+        get_ack_report_computed_task(
+            report_computed_task=report_computed_task
+        ),
+        headers={
+            'Content-Type': 'application/octet-stream',
+        },
+        expected_status=expected_status,
+    )
+
+
+def send_correct_force_get_task_result(
+    cluster_url: str,
+    report_computed_task: ReportComputedTask,
+    expected_status: Optional[int] = None,
+) -> None:
+    response = api_request(
+        cluster_url,
+        'send',
+        REQUESTOR_PRIVATE_KEY,
+        CONCENT_PUBLIC_KEY,
+        get_force_get_task_result(
+            report_computed_task=report_computed_task,
+        ),
+        headers={
+            'Content-Type': 'application/octet-stream',
+        },
+        expected_status=expected_status,
+    )
+    responses_global.append(response.__class__.__name__)
+
+
+@count_fails
+def test_case_multiple_requests_concerning_one_subtask_will_be_processed_one_by_one_if_subtask_exists_in_database(
+    cluster_consts: str,
+    cluster_url: str,
+    task_id: str,
+    subtask_id: str
+) -> None:
+    report_computed_task = create_report_computed_task(task_id=task_id, subtask_id=subtask_id)
+
+    # send messages to save subtask in database and change state to REPORTED. Next message ForceGetTaskResult
+    # requires REPORTED state and this is the simplest way to get it
+    send_correct_force_report_computed_task(
+        cluster_url=cluster_url,
+        report_computed_task=report_computed_task,
+        expected_status=202,
+    )  # Subtask state changed to: FORCING_REPORT
+    send_correct_ack_report_computed_task(
+        cluster_url=cluster_url,
+        report_computed_task=report_computed_task,
+        expected_status=202,
+    )  # Subtask state changed to: REPORTED
+
+    clear_responses()
+    # this is test- sending some messages in one time
+    call_function_in_treads(
+        func=send_correct_force_get_task_result,  # Subtask state changed to: FORCING_RESULT_TRANSFER
+        number_of_threads=NUMBER_OF_TESTING_THREADS,
+        cluster_url=cluster_url,
+        golem_message=report_computed_task,
+    )
+    # waiting for responses from all threads
+    end_time = time.time() + MAXIMUM_WAITING_TIME_FOR_ALL_RESPONSES
+    while len(responses_global) != NUMBER_OF_TESTING_THREADS:
+        time.sleep(0.1)
+        if time.time() >= end_time:
+            break
+
+    expected_responses = ['AckForceGetTaskResult', 'ServiceRefused', 'ServiceRefused']
+    assert len(expected_responses) == NUMBER_OF_TESTING_THREADS,'Did you changed number of testing threads and forgot to change expected_responses?'
+
+    print('Responses = ' + str(responses_global))
+    error_message = f"Responses should be: {expected_responses}."
+    assert compare_lists_regardless_of_order(actual=responses_global, expected=expected_responses), error_message
```

Comment:
this is not OK. I think `assert_condition` should be used here (from `commons`). We cannot use `assert` statement to both
- check if assumptions are correct (sanity check)
- check the result of the test case.
For the latter you have `assertEqual` and similar methods in unit case. Here you don't inherit from `unittest.TestCase` (which is fine, because these are not unit test, but API tests), so a different mechanism has been introduced - `assert_condition` and it should be used in such cases.

Reply:
@pawelkisielewicz, even if so, your function should use the exact mechanism as `assert_condition`, meaning: it should raise `TestAssertionException`, because it is caught in `count_fails` wrapper.

Reply:
Just as @dybi says, both functions should use the same mechanism - i.e. raise the same exceptions on failure.

I did not mean that it was wrong to use `assert_condition()` here but simply that you started adding new, unrelated functionality to it. It's better to have a set of assert functions, each with clearly defined purpose (exact comparison, comparison that ignores order, etc) if the implementations are completely different rather than one, monster function that tries to support every corner case.